### PR TITLE
Support mongo driver 4.x version.

### PR DESCRIPTION
**Pull request by harvies**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/define/MongoDBOperationExecutorInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import net.bytebuddy.matcher.ElementMatchers;
+import org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+import org.apache.skywalking.apm.agent.core.plugin.match.NameMatch;
+
+/**
+ * same whit org.apache.skywalking.apm.plugin.mongodb.v3.define.v38.MongoDBOperationExecutorInstrumentation
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/define/MongoDBOperationExecutorInstrumentation.java:32]:
```suggestion
 * same with org.apache.skywalking.apm.plugin.mongodb.v3.define.v38.MongoDBOperationExecutorInstrumentation
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBClientDelegateInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) {
+        // do nothing
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        Object ret) {
+        if (ret instanceof EnhancedInstance) {
+            // pass remotePeer to OperationExecutor, which will be wrapper as EnhancedInstance
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBClientDelegateInterceptor.java:53]:
```suggestion
            // pass remotePeer to OperationExecutor, which has be enhanced as EnhancedInstance
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBClientDelegateInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            // @see: org.apache.skywalking.apm.plugin.mongodb.v3.define.v37.MongoDBOperationExecutorInstrumentation
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBClientDelegateInterceptor.java:54]:
```suggestion
            // See: org.apache.skywalking.apm.plugin.mongodb.v3.define.v37.MongoDBOperationExecutorInstrumentation
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBOperationExecutorInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+public class MongoDBOperationExecutorInterceptor implements InstanceMethodsAroundInterceptor {
+
+    private static final ILog LOGGER = LogManager.getLogger(MongoDBOperationExecutorInterceptor.class);
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) {
+        String executeMethod = allArguments[0].getClass().getSimpleName();
+        // OperationExecutor has be mark it's remotePeer
+        // @see: MongoDBClientDelegateInterceptor.afterMethod
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBOperationExecutorInterceptor.java:42]:
```suggestion
        // See: MongoDBClientDelegateInterceptor.afterMethod
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBOperationExecutorInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+@SuppressWarnings("Duplicates")
+public class MongoDBOperationExecutorInterceptor implements InstanceMethodsAroundInterceptor {
+
+    private static final ILog LOGGER = LogManager.getLogger(MongoDBOperationExecutorInterceptor.class);
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) {
+        String executeMethod = allArguments[0].getClass().getSimpleName();
+        // OperationExecutor has be mark it's remotePeer
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/interceptor/MongoDBOperationExecutorInterceptor.java:41]:
```suggestion
        // OperationExecutor has included th remotePeer
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/define/MongoDBClientDelegateInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import org.apache.skywalking.apm.plugin.mongodb.v4.interceptor.MongoDBClientDelegateInterceptor;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch.takesArgumentWithType;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+/**
+ * Enhance {@code com.mongodb.client.internal.MongoClientDelegate} instance, and intercept {@code
+ * com.mongodb.client.internal.MongoClientDelegate#getOperationExecutor()}, this is the only way to get
+ * OperationExecutor which is unified entrance of execute mongo command. we can mark OperationExecutor which connection
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/define/MongoDBClientDelegateInstrumentation.java:36]:
```suggestion
 * OperationExecutor which is unified entrance of execute mongo command. Inject the remotePeer into enhanced OperationExecutor.
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/define/MongoDBClientDelegateInstrumentation.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * belongs to.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/define/MongoDBClientDelegateInstrumentation.java:37]:
```suggestion
```

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/support/MongoPluginConfig.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.mongodb.v4.support;
+
+import org.apache.skywalking.apm.agent.core.boot.PluginConfig;
+
+public class MongoPluginConfig {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/support/MongoPluginConfig.java:23]: Is this provided in the 3.x too? With the same config name?

File: test/plugin/scenarios/mongodb-4.x-scenario/bin/startup.sh [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+home="$(cd "$(dirname $0)"; pwd)"
+
+java -jar ${agent_opts} ${home}/../libs/mongodb-4.x-scenario.jar &
```
Comment (wu-sheng) [inline test/plugin/scenarios/mongodb-4.x-scenario/bin/startup.sh:21]: Please activate the parameter collection for testing. And the expected data should be updated too.

**Review by wu-sheng — COMMENTED** [review]
Basically, good to me. Several doc nits. And Test should be enhanced a little.

File: apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/support/MongoPluginConfig.java (hunk 8 above)
Comment (harvies) [reply apm-sniffer/apm-sdk-plugin/mongodb-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/mongodb/v4/support/MongoPluginConfig.java:23]: > Is this provided in the 3.x too? With the same config name?

Yes, the configuration is the same as 3.x.

File: test/plugin/scenarios/mongodb-4.x-scenario/bin/startup.sh (hunk 9 above)
Comment (harvies) [reply test/plugin/scenarios/mongodb-4.x-scenario/bin/startup.sh:21]: > Please activate the parameter collection for testing. And the expected data should be updated too.

I don't quite understand, the integration test has been tested. Can you explain in more detail?

Comment (wu-sheng) [reply test/plugin/scenarios/mongodb-4.x-scenario/bin/startup.sh:21]: The parameter in the MongoPluginConfig. Should be tested in there too.

**Comment by harvies** [comment]
Can you retry the failed CI check for me?
image

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**